transcom/mymove
# MB-12447 Permissions Middleware POC

## [Jira ticket](https://dp3.atlassian.net/browse/MB-12447) for this change

## Summary

This is an initial pass at creating a permissions middleware, building off of previous implementations and research. 

Permissions are at their simplest level a string composed of `action.object` wherein the action is the approved action a user can take on a given object. Actions should follow standard rest CRUD terminology and both actions and objects should align with the endpoint in question. 

Examples:
- `update.move`
- `create.shipment`

Endpoints can have multiple permissions set on them, and a user must have all permissions listed in order to be granted access. Permissions should be defined on the [swagger-definition operation object](https://swagger.io/docs/specification/paths-and-operations/) as an array of `x-permissions` 

A user's role is already determined, so this middleware looks up the permissions already assigned to that role vs. the permissions needed on a given route in order to determine access. Permissions are added onto the session object and LoggedInUserPayload so that the FE can read these. 

At this time the definition of which permissions are assigned to which roles are stored inline at the top of `pkg/auth/authentication/permissions.go`, TBD is a better resting place for the data.

## Testing out this PR ##

- Run migrations, regen swagger files, run the server tests
- Run the server and client, login as TOO to approve a move - everything should go through successfully

## Next steps

- Determine what an appropriate store for permissions should be and update the implementation so that it's no longer hardcoded. Under consideration should be feasibility for maintenance and upkeep as we continue to define permissions in the application. 
- Clean up code base to use permissions middleware instead of checking in-line for roles

## References and Resources

### ADR and previous implementations ###
- [ADR-0040](https://github.com/transcom/mymove/blob/c90ae337459b0dafaede81e8b684f59bf5aa1465/docs/adr/0040-role-base-authorization.md)
-  [Authorization POC](https://github.com/transcom/mymove/pull/2824/files#diff-be793c31d522efea61c667cf49ed11f865eb47751287243daf9d0a4383372eb4) 
- [Permissions POC](https://github.com/transcom/mymove/pull/8578/files#diff-b7c08fec7647d998d5244c3117fee9b0710e74dccb44c7cc1c33fd96cf41b16d) 

### External Resources ###

- [OWASP recommendation for implementing role based access control](https://owasp.org/www-project-proactive-controls/v3/en/c7-enforce-access-controls)
- [Swagger Authentication Documentation](https://swagger.io/docs/specification/2-0/authentication/): Built in support in swagger for auth
- [Swagger composed Auth](https://goswagger.io/tutorial/composed-auth/): how to do more advanced security configurations in Go-Swagger
- [A useful blog post about middleware in general](https://www.alexedwards.net/blog/making-and-using-middleware)
- [Go swagger recommendations for how to use middleware](https://goswagger.io/use/middleware.html)
- [How Go + OpenAPI + Swagger all work together to serve the API](https://eli.thegreenplace.net/2021/rest-servers-in-go-part-4-using-openapi-and-swagger/)
